### PR TITLE
[WIP] enhances persistent connections to use rpc message passing

### PR DESCRIPTION
**Pull request by privateip**
* network_cli now implements connection loca
* network_cli now is a connection plugin
* adds executor/process/connection.py:ConnectionProcess
* moves do_fork() to executor/process/__init__.py
* now checks for instance variable on connection `force_persistent_connection`
* adds utils/jsonrpc.py:JsonRpcServer
* refactors module_utils/connection.py:Connection to accept only socket_path

### CODE REVIEW

File: lib/ansible/executor/process/__init__.py [hunk 1]
```diff
@@ -1,20 +1,45 @@
...
+# (c) 2017, Peter Sprygada <psprygad@redhat.com>
+# (c) 2017 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
 from __future__ import (absolute_import, division, print_function)
 __metaclass__ = type
+
+import os
+import sys
+
+def do_fork():
```
Comment (bcoca) [inline lib/ansible/executor/process/__init__.py]: mostly same as https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/service.py#L94

seems like this is a simpler case

File: lib/ansible/module_utils/service.py [hunk 2]
```diff
@@ -149,6 +140,38 @@ def daemonize(module, cmd):
         if pid > 0:
             os._exit(0)
 
+    return pid
+
+
+def daemonize(module, cmd):
```
Comment (jimi-c) [inline lib/ansible/module_utils/service.py:146]: Why are we doing all this extra boilerplate when multiprocessing.Process does it for us?

File: lib/ansible/executor/process/connection.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+        display.debug('shutdown complete')
+
+    def start(self):
+        """ Called to initiate the connect to the remote device
+        """
+        play_context = self.connection._play_context
+
+        ssh = connection_loader.get('ssh', class_only=True)
+        cp = ssh._create_control_path(play_context.remote_addr, play_context.port, play_context.remote_user)
```
Comment (ganeshrn) [inline lib/ansible/executor/process/connection.py:149]: As some network os support multiple connections type eg. junos support both network_cli and netconf, I think connection name should also be passed to _create_control_path()

File: lib/ansible/executor/process/connection.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+            if os.path.exists(socket_path):
+                elapsed = (C.PERSISTENT_CONNECT_TIMEOUT + 1) - timeout
+                display.vvvv('connected to local socket in %s second(s)' % elapsed, host=play_context.remote_addr)
+                setattr(self.connection, '_connected', True)
+                break
+            time.sleep(1)
+            timeout -= 1
+        else:
+            raise AnsibleError('timeout waiting for local socket', play_context.remote_addr)
+
```
Comment (ganeshrn) [inline lib/ansible/executor/process/connection.py:203]: To handle privilege escalation related activities `update_play_context()` should be called somewhere in this function?

File: lib/ansible/plugins/connection/network_cli.py [hunk 5]
```diff
@@ -68,31 +71,56 @@
...
+
+    def __getattr__(self, name):
+        try:
+            return self.__dict__[name]
+        except KeyError:
+            if name.startswith('_'):
+                raise AttributeError("'%s' object has no attribute '%s'" % (self.__class__.__name__, name))
+            return getattr(self._cliconf, name)
+
+    def exec_command(self, cmd, in_data=None, sudoable=True):
```
Comment (ganeshrn) [inline lib/ansible/plugins/connection/network_cli.py:111]: Should this handle `prompt()`  command to ensure backward compatibility with existing action plugins?

File: lib/ansible/module_utils/service.py (hunk 2 above)
Comment (bcoca) [reply lib/ansible/module_utils/service.py:146]: this is mostly a copy of preexisting code, we should migrate to multiprocessing now that 2.6 is min target. But my original point is still valid, we should have 1 copy of this code.

File: lib/ansible/plugins/connection/network_cli.py (hunk 5 above)
Comment (privateip) [reply lib/ansible/plugins/connection/network_cli.py:111]: this should now be fixed in e4eeebd

**Comment by privateip** [comment]
closing this ... superseded by #31417

**Pull request closed without merging**